pkg/gui/controllers/files_controller.go

```diff
@@ -302,59 +302,128 @@ func (self *FilesController) stageAll() error {
 	return self.contexts.Files.HandleFocus()
 }
 
-func (self *FilesController) ignore(node *filetree.FileNode) error {
-	if node.GetPath() == ".gitignore" {
-		return self.c.ErrorMsg("Cannot ignore .gitignore")
-	}
+func (self *FilesController) unstageFiles(node *filetree.FileNode) error {
 
-	unstageFiles := func() error {
-		return node.ForEachFile(func(file *models.File) error {
-			if file.HasStagedChanges {
-				if err := self.git.WorkingTree.UnStageFile(file.Names(), file.Tracked); err != nil {
-					return err
-				}
+	return node.ForEachFile(func(file *models.File) error {
+		if file.HasStagedChanges {
+			if err := self.git.WorkingTree.UnStageFile(file.Names(), file.Tracked); err != nil {
+				return err
 			}
+		}
 
-			return nil
-		})
-	}
+		return nil
+	})
+}
+
+func (self *FilesController) checkTracking(node *filetree.FileNode, trText string, trPrompt string, trAction string, f func(string) error) (error, bool) {
 
 	if node.GetIsTracked() {
 		return self.c.Confirm(types.ConfirmOpts{
-			Title:  self.c.Tr.IgnoreTracked,
-			Prompt: self.c.Tr.IgnoreTrackedPrompt,
+			Title:  trText,
+			Prompt: trPrompt,
 			HandleConfirm: func() error {
-				self.c.LogAction(self.c.Tr.Actions.IgnoreFile)
+				self.c.LogAction(trAction)
 				// not 100% sure if this is necessary but I'll assume it is
-				if err := unstageFiles(); err != nil {
+				if err := self.unstageFiles(node); err != nil {
 					return err
 				}
 
 				if err := self.git.WorkingTree.RemoveTrackedFiles(node.GetPath()); err != nil {
 					return err
 				}
 
-				if err := self.git.WorkingTree.Ignore(node.GetPath()); err != nil {
+				if err := f(node.GetPath()); err != nil {
 					return err
 				}
+
 				return self.c.Refresh(types.RefreshOptions{Scope: []types.RefreshableView{types.FILES}})
 			},
-		})
+		}), true
 	}
+	return nil, false
+}
 
-	self.c.LogAction(self.c.Tr.Actions.IgnoreFile)
+func (self *FilesController) ignoreOrExcludeFile(node *filetree.FileNode, trText string, trPrompt string, trAction string, f func(string) error) error {
 
-	if err := unstageFiles(); err != nil {
+	trackingErr, earlyExit := self.checkTracking(node, trText, trPrompt, trAction, f)
```

Comment:
now that we only call `checkTracking` from one place, I wanna propose another refactor: there's this idea of the Single Layer of Abstraction where a function should only work with one layer of abstraction, and if you call a function which actually contains code operating on the same layer of abstraction, that could hurt readability.

I propose we do this:
```go
func (self *FilesController) ignoreOrExcludeFile(node *filetree.FileNode, trText string, trPrompt string, trAction string, f func(string) error) error {
	if node.GetIsTracked() {
		return self.c.Confirm(types.ConfirmOpts{
			Title:  trText,
			Prompt: trPrompt,
			HandleConfirm: func() error {
				return self.ignoreOrExcludeTracked(node, trAction, f)
			},
		})
	}

	return self.ignoreOrExcludeUntracked(node, trAction, f)
}
```

`ignoreOrExcludeTracked` and `ignoreOrExcludeUntracked` are operating on the same layer of abstraction (in fact they're very similar functions, so it's good to have them side-by-side where you can see the difference between them).

As for `ignoreOrExcludeFile` it's good to see at a high level that if we're dealing with a tracked file we'll have a confirmation followed by ignoring it, whereas if it's not tracked we'll just go straight for ignoring it.

For what it's worth I considered combining `ignoreOrExcludeTracked` and `ignoreOrExcludeUntracked` into a single function with a bool arg determining whether to run the tracked-file specific code, but in my experience that always ends up being more confusing, even if it reduces the number of lines of code

Reply:
@jesseduffield is done and the integration tests are passing :)